### PR TITLE
feat: permit2 swaps

### PR DESCRIPTION
**Pull request by agualis**
Implements v3 swaps with permit2

v3 swap works as expected (testing with dai-aave and usdc-aave which are the only standard tokens with v3 liquidity after the 10th deployment):
https://mono-test-v3-git-feat-permit2swaps-balancer.vercel.app/swap/sepolia/0xff34b3d4aee8ddcd6f9afffb6fe49bd371b8a357/0x94a9d9ac8a22534e3faca9f4e7f2e2cf85d5e4c8

https://github.com/user-attachments/assets/99e06512-227f-47c6-aab5-ff38efcfa811

v3 pool swap works as expected: 
https://mono-test-v3-git-feat-permit2swaps-balancer.vercel.app/pools/sepolia/v3/0xfd882a11cf794ce28b6c9952192f117cca2d2733/swap

However, some random tx simulation errors were found in the first case with some wallets/accounts. I don't fully understand the root cause yet.

### CODE REVIEW

File: packages/lib/modules/transactions/transaction-steps/useSignPermit2Step.tsx [hunk 1]
```diff
@@ -62,25 +64,25 @@ export function useSignPermit2Step(params: AddLiquidityPermit2Params): Transacti
...
       <VStack width="full">
-        {error && <BalAlert content={error} status="error" />}
+        {error ? <BalAlert content={error} status="error" /> : null}
         {!isConnected && <ConnectWallet isLoading={isLoading} width="full" />}
-        {shouldChangeNetwork && isConnected && (
+        {shouldChangeNetwork && isConnected ? (
           <NetworkSwitchButton {...networkSwitchButtonProps} />
-        )}
-        {!shouldChangeNetwork && isConnected && (
+        ) : null}
```
Comment (garethfuller) [inline packages/lib/modules/transactions/transaction-steps/useSignPermit2Step.tsx:71]: Why ? then null instead of just &&?

Comment (agualis) [reply packages/lib/modules/transactions/transaction-steps/useSignPermit2Step.tsx]: Fixed. I did this during the monorepo + eslint migration and somehow my lint auto formatted did this by mistake.

**Comment by garethfuller** [comment]
@agualis this first link in the description leads to a full page error for me.

**Review by garethfuller — CHANGES_REQUESTED** [review]
I can't select the tokens.

[https://www.loom.com/share/4cd09c325b7c48a2b1c4998cd22c8861?sid=baca4be7-
1f8f-476d-a8bd-5221434a4b8d](https://www.loom.com/share/4cd09c325b7c48a2b1c4998cd22c8861?sid=baca4be7-1f8f-476d-a8bd-5221434a4b8d)

**Comment by agualis** [comment]
> @agualis this first link in the description leads to a full page error for me.

Updated the link

**Comment by agualis** [comment]
> I can't select the tokens.
> 
> [https://www.loom.com/share/4cd09c325b7c48a2b1c4998cd22c8861?sid=baca4be7- 1f8f-476d-a8bd-5221434a4b8d](https://www.loom.com/share/4cd09c325b7c48a2b1c4998cd22c8861?sid=baca4be7-1f8f-476d-a8bd-5221434a4b8d)

This is fixed.

File: packages/lib/modules/tokens/approvals/permit2/permit2.helpers.ts [hunk 2]
```diff
@@ -1,64 +1,94 @@
...
-    amountsIn: queryOutput.sdkQueryOutput.amountsIn,
-  }).map(a => a.token.address)
-  return result
+    tokenAmountsIn,
+  }).map(t => t.address)
+}
+
+export function permit2Address(chain: GqlChain): Address {
+  return getNetworkConfig(chain).contracts.permit2 || ('' as Address)
+}
```
Comment (garethfuller) [inline packages/lib/modules/tokens/approvals/permit2/permit2.helpers.ts:80]: Should this really return an empty address ever?

Comment (agualis) [reply packages/lib/modules/tokens/approvals/permit2/permit2.helpers.ts]: Added a TODO comment before that line:
`  // TODO: Remove ('' as Address) when all chains have permit2 defined to it's not optional anymore`

File: packages/lib/modules/pool/actions/add-liquidity/handlers/BaseUnbalancedAddLiquidity.handler.ts [hunk 3]
```diff
@@ -66,6 +69,7 @@ export abstract class BaseUnbalancedAddLiquidityHandler implements AddLiquidityH
       rpcUrl: getRpcUrl(this.helpers.chainId),
       amountsIn,
       kind: AddLiquidityKind.Unbalanced,
+      userData,
```
Comment (garethfuller) [inline packages/lib/modules/pool/actions/add-liquidity/handlers/BaseUnbalancedAddLiquidity.handler.ts:72]: Not sure why this is called userData if it's just the address?

Comment (agualis) [reply packages/lib/modules/pool/actions/add-liquidity/handlers/BaseUnbalancedAddLiquidity.handler.ts:72]: Good catch, this was a typo. `userData` will be used by hooks in incoming integrations (not yet).
Replaced with `sender`.

**Review by garethfuller — CHANGES_REQUESTED** [review]
Tried to do a v2 swap on gnosis.
Screenshot 2024-11-05 at 15 22 56

**Comment by agualis** [comment]
> Tried to do a v2 swap on gnosis. 
Fixed after SDK fix + bump.

**Review by garethfuller — APPROVED** [review]

**Pull request merged**